server/event.go

```diff
@@ -0,0 +1,107 @@
+package server
+
+import (
+	"fmt"
+	"github.com/Kong/go-pdk"
+	"time"
+)
+
+// Incoming data for a new event.
+// TODO: add some relevant data to reduce number of callbacks.
+type StartEventData struct {
+	InstanceId int    // Instance ID to start the event
+	EventName  string // event name (not handler method name)
+	// ....
+}
+
+type eventData struct {
+	id       int              // event id
+	instance *instanceData    // plugin instance
+	ipc      chan interface{} // communication channel (TODO: use decoded structs)
+	pdk      *pdk.PDK         // go-pdk instance
+}
+
+func (rh *rpcHandler) addEvent(event *eventData) {
+	rh.lock.Lock()
+	defer rh.lock.Unlock()
+
+	event.id = rh.nextEventId
+	rh.nextEventId++
+	rh.events[event.id] = event
+}
+
+// HandleEvent starts the call/{callback/response}*/finish cycle.
+// More than one event can be run concurrenty for a single plugin instance,
+// they all receive the same object instance, so should be careful if it's
+// mutated or holds references to mutable data.
+//
+// RPC exported method
+func (rh *rpcHandler) HandleEvent(in StartEventData, out *StepData) error {
+	rh.lock.RLock()
+	instance, ok := rh.instances[in.InstanceId]
+	rh.lock.RUnlock()
+	if !ok {
+		return fmt.Errorf("No plugin instance %d", in.InstanceId)
+	}
+
+	h, ok := instance.handlers[in.EventName]
+	if !ok {
+		return fmt.Errorf("undefined method %s", in.EventName)
+	}
+
+	ipc := make(chan interface{})
+
+	event := eventData{
+		instance: instance,
+		ipc:      ipc,
+		pdk:      pdk.Init(ipc),
+	}
+
+	rh.addEvent(&event)
+
+	//log.Printf("Will launch goroutine for key %d / operation %s\n", key, op)
+	go func() {
+		_ = <-ipc
+		h(event.pdk)
+
+		func() {
+			defer func() { recover() }()
+			ipc <- "ret"
+		}()
```

Comment:
yes.  i tried hard to find some less verbose way to do it, but failed miserably.  I guess it's intentional as a way to make it easier to handle errors than ignore them.

in this case it's needed because the channel can be closed from the other side (when calling exit() from the plugin), and there's no "try to send" operation.  if the channel is closed, you get a hard error.  this is the documented way to do it.